### PR TITLE
[FIX] Fix quote removal coming from variable expansions

### LINKED ISSUE #173: [BUG] Quotes that came from variable expansions should not be removed
#### Test case:
```sh
export S="'"
$S""$S
```
Bash:
```sh
bash-5.1$ echo $S""$S
bash: ''
```
🌊rash:
```sh
🌊rash$ echo $S""$S
🌊rash: ""
```

> After these expansions are performed, quote characters present **in the original word** are removed unless they have been quoted themselves (quote removal).
https://www.gnu.org/savannah-checkouts/gnu/bash/manual/html_node/Shell-Expansions.html

> After the preceding expansions, all unquoted occurrences of the characters ‘\’, ‘'’, and ‘"’ **that did not result from one of the above expansions** are removed.
https://www.gnu.org/savannah-checkouts/gnu/bash/manual/html_node/Quote-Removal.html

---
Quote removal **has** to be done after parameter expansion, otherwise this test case would become an issue:
```sh
echo $US""ER
ER
```

---
- [ ] Also test with `${` in variables.

### PR DESCRIPTION
**Pull request by itislu**
* Instead of immediatly expanding into the string, create a stack of tasks that mark the index and the length of one section important for the expander.

* A stack, bc the expander has to finally expand the sequences from the back of the string to the front.
This way, the indexes don't have to be updated after one section of the string has been expanded.

* The issue of the removal of quotes coming from variables gets fixed bc at time of creation of the tasks, the string to be expanded does not get changed yet, and so the contents of the variables cannot disrupt.

---
### Why the regression test fails:
- `$"<string>"` (locale translation) and 
- `$'<string>'` (escape sequence) ...

... have a special meaning in bash, that's why the dollar sign and the quotes get removed.

  Previously I just removed the `$`, but I didn't fully understand why it should be this way.
  Now that I know, I decided to not do that and handle the `$` like every other character in such a case, and therefore keep the dollar sign, bc we don't support these features (locale translation and escape sequence). 

  I manually tested all failed test cases of the tester, and in all cases the only difference to bash is that we don't remove the dollar sign.
- [ ]  @LeaYeh What do you think?

**Bash manual references:**
[Dollar single quote](https://www.gnu.org/software/bash/manual/html_node/ANSI_002dC-Quoting.html)
[Dollar double quote](https://www.gnu.org/software/bash/manual/html_node/Locale-Translation.html)

---
* Resolves #173

### CODE REVIEW

File: source/frontend/expander/quote_removal.c
Comment (itislu) [inline source/frontend/expander/quote_removal.c:1]: With the new expander framework, the core functions that do the actual expansion got very simple.
All the data they need is already saved in the `expander_task` struct.

File: source/frontend/expander/special_param_expansion.c
Comment (itislu) [inline source/frontend/expander/special_param_expansion.c:1]: Same here, no more logic required.

File: source/frontend/expander/variable_expansion.c
Comment (itislu) [inline source/frontend/expander/variable_expansion.c:1]: The only logci left here is that if a variable was not found in the environment, it still has to replace the `$VARNAME`.
This is done by giving an empty string `""` as the replacement argument to `ft_strrplc_part()`.

File: source/utils/string_utils.c [hunk 1]
```diff
@@ -24,6 +24,39 @@ bool	replace_string_content(char **str, char *new_content)
...
+	if (operation == OP_CLEAN)
+		return (0);
+	if (c == '\'')
+		return (0);
+	if (c == '"')
+		return (1);
+	return (-1);
+}
+
+bool	is_open_pair(unsigned char c, t_is_open_pair_op operation)
```
Comment (itislu) [inline source/utils/string_utils.c:38]: Just moved from expander_utils.c to string_utils.c

File: source/frontend/expander/expansion_handler.c [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+			t_expander_op op_mask)
+{
+	t_list	*task_stack;
+
+	task_stack = NULL;
+	if (!create_expander_task_stack(&task_stack, *new_str, op_mask) || \
+		!execute_expander_task_stack(new_str, task_stack, lst, shell))
+		return (ft_lstclear(&task_stack, (void *)free_expander_task), false);
+	if (is_null_expansion(*new_str, task_stack))
+		ft_free_and_null((void **)new_str);
```
Comment (itislu) [inline source/frontend/expander/expansion_handler.c:13]: If nothing but an empty string `""` is left after expansion, make it `NULL`. Unless it was previously in quotes.
The `is_null_expansion()` function goes through the whole task_stack and checks for that, i.e. if there was any quote removal performed.

File: source/frontend/expander/expansion_handler.c [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	ft_lstclear(&task_stack, (void *)free_expander_task);
+	return (true);
+}
+
+bool	execute_expander_task_stack(char **new_str, t_list *task_stack,
+			t_list **lst, t_shell *shell)
```
Comment (itislu) [inline source/frontend/expander/expansion_handler.c:19]: Modifies the `new_str` at the saved indexes provided by the `task_stack`. The characters not relevant for the expander stay untouched.

File: source/utils/string_utils.c [hunk 4]
```diff
@@ -103,6 +136,8 @@ bool	skip_single_quote(char *str, size_t *i)
 {
 	size_t	start;
 
+	if (is_open_pair('\'', OP_GET))
+		return (false);
```
Comment (itislu) [inline source/utils/string_utils.c:140]: If this function gets called and it is already inside of a single quote, it returns false to indicate that the end of a single quote pair has is already here and that the index has not been moved.

File: source/frontend/expander/bad_substitution.c [hunk 5]
```diff
@@ -53,10 +53,12 @@ bool	is_bad_substitution(char *str, t_expander_op op_mask)
 {
 	size_t	i;
 
+	if (!(op_mask & (E_EXPAND | E_HEREDOC)))
+		return (false);
```
Comment (itislu) [inline source/frontend/expander/bad_substitution.c:57]: If no variable expansion will be performed, bad substitution is not possible and should therefor not be checked.

**Comment by LeaYeh** [comment]
For me as a project developer and the corner case is not clearly declared in the subject so we have the right to design our own shell behavior. Doing the evaluation we have a very clearly defense why do we decide to design in these way with a make sense reason that is totally enough.

**Comment by LeaYeh** [comment]
In my mind, we are not going to be trained as a coding monkey but trained ourselves to have a higher view like a researcher and designer is much more important

**Review by LeaYeh — APPROVED** [review]

**Pull request merged**